base-R.qmd

```diff
@@ -0,0 +1,502 @@
+
+Here are a couple of examples:
+
+```{r}
+df <- tibble(x = 1:3, y = c("a", "e", "f"), z = runif(3))
+
+# Select first row and second column
+df[1, 2]
+
+# Select all rows and columns x and y
+df[, c("x" , "y")]
+
+# Select rows where `x` is greater than 1 and all columns
+df[df$x > 1, ]
+```
+
+We'll come back to `$` shortly, but we need to use it here because `[` doesn't use tidy evaluation, so you need to be explicit about the source of the `x` variable.
+
+There's an important difference between tibbles and data frames when it comes to `[`.
+If `df` is a data frame, `df[, cols]` will return a vector if `col` selects a single column and a data frame if it selects more than one column.
+If `df` is a tibble, then `[` will always return a tibble.
+
+```{r}
+df1 <- data.frame(x = 1:3)
+df1[, "x"]
+
+df2 <- tibble(x = 1:3)
+df2[, "x"]
+```
+
+You can also avoid this problem by using 1d `[` to select columns:
+
+```{r}
+df1["x"]
+```
+
+### dplyr equivalents
+
+A number of dplyr verbs are special cases of `[`:
+
+-   `filter()` is equivalent to subsetting the rows with a logical vector, taking care to exclude missing values:
+
+    ```{r}
+    #| results: false
+    df <- tibble(x = c(2, 3, 1, 1, NA), y = letters[1:5], z = runif(5))
+    df |> filter(x > 1)
+
+    # same as
+    df[!is.na(df$x) & df$x > 1, ]
+    ```
+
+    Another common technique in the wild is to `which()` for it's side-effect of dropping missing values: `df[which(df$x > 1), ]`
+
+-   `arrange()` is equivalent to subsetting the rows with an integer vector, usually created with `order()`.
+
+    ```{r}
+    #| results: false
+    df |> arrange(x, y)
+
+    # same as
+    df[order(df$x, df$y), ]
+    ```
+
+    You can use the `decreasing` argument to sort all columns in decreasing order or `-rank(col)` to individual sort column in decreasing order.
+
+-   Both `select()` and `relocate()` are similar to subsetting the columns with a character vector.
+
+    ```{r}
+    #| results: false
+    df |> select(x, z)
+    # same as
+    df[, c("x", "z")]
+    ```
+
+base R also provides a function that combines the features of `filter()` and `select()`[^base-r-2] called `subset()`:
+
+[^base-r-2]: But it doesn't handle grouped data frames differently and it doesn't support the more selection helper functions like `starts_with()`.
+
+```{r}
+df |> 
+  filter(x > 1) |> 
+  select(y, z)
+
+df |> subset(x > 1, c(y, z))
+```
+
+This function was the inspiration for much of dplyr's syntax.
+
+### Exercises
+
+1.  Create functions that take a vector as input and return:
+
+    a.  The elements at even numbered positions.
+    b.  Every element except the last value.
+    c.  Only even numbers (and no missing values).
+
+2.  Why is `x[-which(x > 0)]` not the same as `x[x <= 0]`?
+    Read the documentation for `which()` and do some experiments to figure it out.
+
```

Comment:
Additional exercise idea:

Suppose you have the following data frame:

```r
df <- tibble(x = c(1, 2, NA, 3, 4, 5, NA))
```

Subset for rows where `x > 2` using `filter()` and also using `[`. How do the results compare?